File: packages/railtracks/src/railtracks/llm/tools/parameters/object_parameter.py
from typing import Any, Dict, Optional

from ._base import Parameter, ParameterType


class ObjectParameter(Parameter):
    param_type: ParameterType = ParameterType.OBJECT

    def __init__(
        self,
        name: str,
        properties: list[Parameter],
        description: Optional[str] = None,
        required: bool = True,
        additional_properties: bool = False,
        default: Any = None,
    ):
        super().__init__(name, description, required, default)
        self.properties = properties
        self.additional_properties = additional_properties

    def to_json_schema(self) -> Dict[str, Any]:
        schema = {
            "type": "object",
            "properties": {},
            "additionalProperties": self.additional_properties,
        }

        if self.description:
            schema["description"] = self.description

        required_props = []
        for prop in self.properties:
            schema["properties"][prop.name] = prop.to_json_schema()
            if prop.required:
                required_props.append(prop.name)

        if required_props:
            schema["required"] = required_props

        if self.default is not None:
            schema["default"] = self.default

        if self.enum:
            schema["enum"] = self.enum

        return schema

    def __repr__(self) -> str:
        return (
            f"ObjectParameter(name={self.name!r}, properties={self.properties!r}, "
            f"description={self.description!r}, required={self.required!r}, "
            f"additional_properties={self.additional_properties!r}, default={self.default!r})"
        )

File: dotnet/src/Railtracks.Core/Llm/Tools/ObjectParameter.cs
namespace Railtracks.Llm.Tools;

/// <summary>Parameter representing an object type. Ported from <c>ObjectParameter</c>.</summary>
public sealed class ObjectParameter : Parameter
{
    public ObjectParameter(
        string name,
        IReadOnlyList<Parameter> properties,
        string? description = null,
        bool required = true,
        bool additionalProperties = false,
        object? @default = null)
        : base(name, description, required, @default)
    {
        Properties = properties;
        AdditionalProperties = additionalProperties;
        ParamType = ParameterType.Object.Value();
    }

    public IReadOnlyList<Parameter> Properties { get; }
    public bool AdditionalProperties { get; }

    public override Dictionary<string, object?> ToJsonSchema()
    {
        var properties = new Dictionary<string, object?>();
        var requiredProps = new List<string>();

        foreach (var prop in Properties)
        {
            properties[prop.Name] = prop.ToJsonSchema();
            if (prop.Required)
                requiredProps.Add(prop.Name);
        }

        var schema = new Dictionary<string, object?>
        {
            ["type"] = "object",
            ["properties"] = properties,
            ["additionalProperties"] = AdditionalProperties,
        };

        if (!string.IsNullOrEmpty(Description))
            schema["description"] = Description;
        if (requiredProps.Count > 0)
            schema["required"] = requiredProps;
        if (Default is not null)
            schema["default"] = Default;
        if (Enum is { Count: > 0 })
            schema["enum"] = Enum;

        return schema;
    }

    public override string ToString() =>
        $"ObjectParameter(name='{Name}', properties=[{string.Join(", ", Properties)}], " +
        $"description='{Description}', required={Required}, " +
        $"additional_properties={AdditionalProperties}, default={Default})";
}
